### PR TITLE
Implemented descriptor wrappers in Python

### PR DESCRIPTION
**Pull request by haberman**

### CODE REVIEW

File: python/descriptor.c [hunk 1]
```diff
@@ -38,113 +42,919 @@
...
+}
+
+typedef void* PyUpb_ToProto_Func(const void* def, upb_arena* arena);
+
+static PyObject* PyUpb_DescriptorBase_CopyToProto(PyObject* _self,
+                                                  PyUpb_ToProto_Func* func,
+                                                  const upb_msglayout* layout,
+                                                  PyObject* py_proto) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  upb_arena* arena = upb_arena_new();
```
Comment (fowles) [inline python/descriptor.c:152]: can arena_new fail?

File: python/descriptor.c [hunk 2]
```diff
@@ -38,113 +42,919 @@
...
   Py_DECREF(base->pool);
-  PyObject_Del(self);
+  Py_XDECREF(base->options);
+  PyUpb_Dealloc(base);
+}
+
+#define DESCRIPTOR_BASE_SLOTS                           \
+  {Py_tp_new, (void*)&PyUpb_Forbidden_New}, {           \
+    Py_tp_dealloc, (void*)&PyUpb_DescriptorBase_Dealloc \
+  }
```
Comment (fowles) [inline python/descriptor.c:174]: the formatting here feel a bit offf.  did clang-format really do this?  can you force a new line between the two elements?

File: python/descriptor.c [hunk 3]
```diff
@@ -38,113 +42,919 @@
...
+      upb_symtab_lookupenum(symtab, PyUnicode_AsUTF8AndSize(qname, NULL));
+  Py_DECREF(qname);
+  return ret;
+}
+
+static const void* PyUpb_Descriptor_LookupNestedExtension(const upb_msgdef* m,
+                                                          const char* name) {
+  const upb_filedef* filedef = upb_msgdef_file(m);
+  const upb_symtab* symtab = upb_filedef_symtab(filedef);
+  PyObject* qname = PyUnicode_FromFormat("%s.%s", upb_msgdef_fullname(m), name);
```
Comment (fowles) [inline python/descriptor.c:223]: why do you use PyUnicode_FromFormat instead of something lower level like sprintf?

File: python/descriptor.c [hunk 4]
```diff
@@ -38,113 +42,919 @@
...
+      _self, (PyUpb_ToProto_Func*)&upb_MessageDef_ToProto,
+      &google_protobuf_DescriptorProto_msginit, py_proto);
+}
+
+static PyObject* PyUpb_Descriptor_EnumValueName(PyObject* _self,
+                                                PyObject* args) {
+  PyUpb_DescriptorBase* self = (void*)_self;
+  const char* enum_name;
+  int number;
+  if (!PyArg_ParseTuple(args, "si", &enum_name, &number)) return NULL;
```
Comment (fowles) [inline python/descriptor.c:312]: should you call some kind of PyErr on failure here?

File: python/descriptor.c [hunk 5]
```diff
@@ -38,113 +42,919 @@
...
+          (void*)&upb_msgdef_field,
+          (void*)&PyUpb_FieldDescriptor_Get,
+      },
+      (void*)&upb_msgdef_ntofz,
+      (void*)&upb_fielddef_name,
+  };
+  return PyUpb_ByNameMap_New(&funcs, self->def, self->pool);
+}
+
+static PyObject* PyUpb_Descriptor_GetFieldsByCamelcaseName(PyObject* _self,
```
Comment (fowles) [inline python/descriptor.c:342]: CamelCase (I think)

File: python/descriptor.c [hunk 6]
```diff
@@ -38,113 +42,919 @@
...
+     "Nested types by name"},
+    {"extensions", PyUpb_Descriptor_GetExtensions, NULL, "Extensions Sequence"},
+    {"extensions_by_name", PyUpb_Descriptor_GetExtensionsByName, NULL,
+     "Extensions by name"},
+    {"extension_ranges", PyUpb_Descriptor_GetExtensionRanges, NULL,
+     "Extension ranges"},
+    {"enum_types", PyUpb_Descriptor_GetEnumTypes, NULL, "Enum sequence"},
+    {"enum_types_by_name", PyUpb_Descriptor_GetEnumTypesByName, NULL,
+     "Enum types by name"},
+    //{ "enum_values_by_name", PyUpb_Descriptor_GetEnumValuesByName, NULL,
```
Comment (fowles) [inline python/descriptor.c:524]: commented out code

File: python/descriptor.c [hunk 7]
```diff
@@ -38,113 +42,919 @@
...
+        "GetOptions",
+        PyUpb_EnumDescriptor_GetOptions,
+        METH_NOARGS,
+    },
+    {
+        "CopyToProto",
+        PyUpb_EnumDescriptor_CopyToProto,
+        METH_O,
+    },
+    {NULL}};
```
Comment (fowles) [inline python/descriptor.c:686]: why does this one have so much more whitespace?

File: python/descriptor.c [hunk 8]
```diff
@@ -38,113 +42,919 @@
...
+  return PyUpb_DescriptorBase_GetOptions(
+      self, upb_enumvaldef_options(self->def),
+      &google_protobuf_EnumValueOptions_msginit,
+      "google.protobuf.EnumValueOptions");
+}
+
+static PyGetSetDef PyUpb_EnumValueDescriptor_Getters[] = {
+    {"name", PyUpb_EnumValueDescriptor_GetName, NULL, "name"},
+    {"number", PyUpb_EnumValueDescriptor_GetNumber, NULL, "number"},
+    //{ "index", (getter)GetIndex, NULL, "index"},
```
Comment (fowles) [inline python/descriptor.c:747]: commented out code

File: python/descriptor.c [hunk 9]
```diff
@@ -38,113 +42,919 @@
...
+  return PyUnicode_FromString(upb_fielddef_jsonname(self->def));
+}
+
+static PyObject* PyUpb_FieldDescriptor_GetFile(PyUpb_DescriptorBase* self,
+                                               void* closure) {
+  const upb_filedef* file = upb_fielddef_file(self->def);
+  if (file) {
+    return PyUpb_FileDescriptor_Get(file);
+  } else {
+    Py_RETURN_NONE;
```
Comment (fowles) [inline python/descriptor.c:817]: I generally prefer the structure of

```
if (!file) Py_RETURN_NONE;

return ...
```

as that follows the common idiom of bailing from error early and having the mainline continue, so the code requires less churn if it evolves in the future

File: python/descriptor.c [hunk 10]
```diff
@@ -38,113 +42,919 @@
...
+static PyObject* PyUpb_FieldDescriptor_GetType(PyUpb_DescriptorBase* self,
+                                               void* closure) {
   return PyLong_FromLong(upb_fielddef_descriptortype(self->def));
 }
 
-static PyObject *PyUpb_FieldDescriptor_GetLabel(PyUpb_DescriptorBase *self,
-                                                void *closure) {
+static PyObject* PyUpb_FieldDescriptor_GetCppType(PyUpb_DescriptorBase* self,
+                                                  void* closure) {
+  static const uint8_t cpp_types[] = {-1, 7, 6, 1, 3, 8, 10, 5, 2, 4, 9, 9};
```
Comment (fowles) [inline python/descriptor.c:828]: some commment about where these constants come from

File: python/descriptor.c [hunk 11]
```diff
@@ -38,113 +42,919 @@
...
-    (getter)GetHasOptions, (setter)SetHasOptions, "Has Options"}, { "_options",
-    (getter)NULL, (setter)SetOptions, "Options"}, { "_serialized_options",
-    (getter)NULL, (setter)SetSerializedOptions, "Serialized Options"},
-  */
+    {"index", (getter)PyUpb_FieldDescriptor_GetIndex, NULL, "Index"},
+    {"default_value", (getter)PyUpb_FieldDescriptor_GetDefaultValue, NULL,
+     "Default Value"},
+    {"has_default_value", (getter)PyUpb_FieldDescriptor_HasDefaultValue},
+    {"is_extension", (getter)PyUpb_FieldDescriptor_GetIsExtension, NULL, "ID"},
+    //{ "id", (getter)GetID, NULL, "ID"},
```
Comment (fowles) [inline python/descriptor.c:935]: a bunch more commented out bits

File: python/descriptor.c [hunk 12]
```diff
@@ -224,29 +1193,326 @@ static PyType_Spec PyUpb_FileDescriptor_Spec = {
...
+                                                    PyObject* py_proto) {
+  return PyUpb_DescriptorBase_CopyToProto(
+      _self, (PyUpb_ToProto_Func*)&upb_MethodDef_ToProto,
+      &google_protobuf_MethodDescriptorProto_msginit, py_proto);
+}
+
+static PyGetSetDef PyUpb_MethodDescriptor_Getters[] = {
+    {"name", PyUpb_MethodDescriptor_GetName, NULL, "Name", NULL},
+    {"full_name", PyUpb_MethodDescriptor_GetFullName, NULL, "Full name", NULL},
+    //{ "index", PyUpb_MethodDescriptor_GetIndex, NULL, "Index", NULL},
```
Comment (fowles) [inline python/descriptor.c:1264]: commented out code

File: python/descriptor.c [hunk 13]
```diff
@@ -224,29 +1193,326 @@ static PyType_Spec PyUpb_FileDescriptor_Spec = {
...
+}
+
+static PyGetSetDef PyUpb_OneofDescriptor_Getters[] = {
+    {"name", PyUpb_OneofDescriptor_GetName, NULL, "Name"},
+    {"full_name", PyUpb_OneofDescriptor_GetFullName, NULL, "Full name"},
+    {"index", PyUpb_OneofDescriptor_GetIndex, NULL, "Index"},
+    {"containing_type", PyUpb_OneofDescriptor_GetContainingType, NULL,
+     "Containing type"},
+    {"has_options", PyUpb_OneofDescriptor_GetHasOptions, NULL, "Has Options"},
+    //{ "fields", (getter)GetFields, NULL, "Fields"},
```
Comment (fowles) [inline python/descriptor.c:1352]: here too

File: python/descriptor_pool.c [hunk 14]
```diff
@@ -62,10 +72,17 @@ static int PyUpb_DescriptorPool_Clear(PyUpb_DescriptorPool* self) {
   return 0;
 }
 
+PyObject* PyUpb_DescriptorPool_Get(const upb_symtab *symtab) {
+  PyObject *pool = PyUpb_ObjCache_Get(symtab);
```
Comment (fowles) [inline python/descriptor_pool.c:76]: you got some pointers wandering from east to west, run clang-format

File: python/protobuf.c [hunk 15]
```diff
@@ -90,6 +91,49 @@ PyObject *PyUpb_ObjCache_Get(const void *key) {
   }
 }
 
+// -----------------------------------------------------------------------------
+// Arena
+// -----------------------------------------------------------------------------
+
+typedef struct {
+  PyObject_HEAD upb_arena* arena;
```
Comment (fowles) [inline python/protobuf.c:99]: is there supposed to be a `;` after HEAD?

File: python/protobuf.h [hunk 16]
```diff
@@ -104,6 +107,8 @@ PyTypeObject *AddObject(PyObject *m, const char *name, PyType_Spec *spec);
 // Creates a Python type from `spec` and adds it to the given module `m`.
 PyTypeObject *PyUpb_AddClass(PyObject *m, PyType_Spec *spec);
 
+PyObject *PyUpb_Forbidden_New(PyObject *cls, PyObject *args, PyObject *kwds);
```
Comment (fowles) [inline python/protobuf.h:110]: function level comment please

File: python/descriptor.c [hunk 17]
```diff
@@ -38,113 +42,919 @@
...
+
+static PyObject* PyUpb_DescriptorBase_GetOptions(PyUpb_DescriptorBase* self,
+                                                 const upb_msg* opts,
+                                                 const upb_msglayout* layout,
+                                                 const char* msg_name) {
+  if (!self->options) {
+    // Load descriptors protos if they are not loaded already. We have to do
+    // this lazily, otherwise, it would lead to circular imports.
+    PyObject* mod = PyImport_ImportModule("google.protobuf.descriptor_pb2");
+    Py_DECREF(mod);
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/descriptor.c:110]: We don't need a ref to it, we only need to ensure that it has been imported and the respective classes have been defined in the global namespace.

File: python/descriptor.c (hunk 1 above)
Comment (haberman) [reply python/descriptor.c:152]: Done.

File: python/descriptor.c (hunk 3 above)
Comment (haberman) [reply python/descriptor.c:223]: Yes `clang-format` did this. Fixed manually.

File: python/descriptor.c (hunk 4 above)
Comment (haberman) [reply python/descriptor.c:312]: No, PyArg_ParseTuple() handles this.

File: python/descriptor.c (hunk 5 above)
Comment (haberman) [reply python/descriptor.c:342]: Done.

File: python/descriptor.c (hunk 6 above)
Comment (haberman) [reply python/descriptor.c:524]: Done.

File: python/descriptor.c (hunk 7 above)
Comment (haberman) [reply python/descriptor.c:686]: I think it was because there was a trailing comma. Fixed.

File: python/descriptor.c (hunk 8 above)
Comment (haberman) [reply python/descriptor.c:747]: Added TODO.

File: python/descriptor.c (hunk 9 above)
Comment (haberman) [reply python/descriptor.c:817]: Done.

File: python/descriptor.c (hunk 11 above)
Comment (haberman) [reply python/descriptor.c:935]: Done.

File: python/descriptor.c (hunk 12 above)
Comment (haberman) [reply python/descriptor.c:1264]: Done.

File: python/descriptor.c (hunk 13 above)
Comment (haberman) [reply python/descriptor.c:1352]: Done.

File: python/protobuf.c (hunk 15 above)
Comment (haberman) [reply python/protobuf.c:99]: No, unfortunately it's a macro that already specifies the semicolon. :(  This causes the bad clang-format, and I didn't find a good solution to make clang-format do something reasonable here. Fixed manually.

File: python/protobuf.h (hunk 16 above)
Comment (haberman) [reply python/protobuf.h:110]: Done.

File: python/descriptor_pool.c (hunk 14 above)
Comment (haberman) [reply python/descriptor_pool.c]: Done.

File: python/descriptor.c (hunk 10 above)
Comment (haberman) [reply python/descriptor.c]: Done.

**Comment by haberman** [comment]
PTAL.

File: python/descriptor_pool.c [hunk 18]
```diff
@@ -117,6 +117,7 @@ static PyObject* PyUpb_DescriptorPool_AddSerializedFile(
   char* buf;
   Py_ssize_t size;
   upb_arena* arena = upb_arena_new();
+  if (!arena) PYUPB_RETURN_OOM;
   PyObject* result = NULL;
 
   if (self->db) {
```
Comment (fowles) [inline python/descriptor_pool.c:123]: missing arena free on this path

File: python/protobuf.c [hunk 19]
```diff
@@ -96,7 +96,8 @@ PyObject *PyUpb_ObjCache_Get(const void *key) {
 // -----------------------------------------------------------------------------
 
 typedef struct {
-  PyObject_HEAD upb_arena* arena;
+  PyObject_HEAD
```
Comment (fowles) [inline python/protobuf.c:99]: I suspect you can add a // at the end of the line to force clang-format to line break appropriately

Comment (haberman) [reply python/protobuf.c:99]: I tried that, and it does prevent the line join, but it indents the next line:

```c
typedef struct {
  PyObject_HEAD  // Macro.
      upb_arena* arena;
} PyUpb_Arena;
```

File: python/descriptor_pool.c (hunk 18 above)
Comment (haberman) [reply python/descriptor_pool.c:123]: Done.

**Comment by haberman** [comment]
PTAL.

**Review by fowles — APPROVED** [review]

**Pull request merged**